### ISSUE TITLE
Help explain "Actually better for Inductor to codegen attention here"

### ISSUE THREAD

**Issue by huntzhan** [pytorch-labs/gpt-fast#31]
Can you help explain this comment? What is the best setup for `torch.compile`?
 
https://github.com/pytorch-labs/gpt-fast/blob/db7b273ab86b75358bd3b014f1f022a19aba4797/generate.py#L64-L67

**Comment by Chillee** [comment]
`F.scaled_dot_product_attention` automatically makes a decision about what backend to dispatch to. For example, it can choose to dispatch to the FlashAttention2 kernel. Or, for example, on platforms where FlashAttention2 is not supported, it can choose to dispatch to the "math" implementation, which is attention implemented using more primitive PyTorch operators.

In the case of decoding, however, the FlashAttention algorithm is *not* beneficial. In fact, it's actively detrimental. So in this case, it's better to dispatch to more primitive operators, where `torch.compile` can codegen the kernels from scratch.

**Comment by huntzhan** [comment]
Thank a lot!

**Comment by Chillee** [comment]
> The function decode_n_tokens, in which the torch.backends.cuda.sdp_kernel decorator is used, is not compiled. Does that mean the aforementioned behavior is not applied?

No, `decode_n_tokens` calls `decode_token`, which does have the decorator. And the annotation is still respected there.

**Comment by huntzhan** [comment]
> Inspired by your project, I've successfully applied the optimization strategy to baichuan 13b: https://github.com/armed-gpt/gpt-blazing Could I submit a PR to add a reference case to the README?

Hi @Chillee, I've submitted a PR #48, could you make a comment about it? thanks.
